### DISCUSSION TITLE
How to access the `state` inside the `.forEach`?

### DISCUSSION THREAD

**Discussion by MojtabaHs** [pointfreeco/swift-composable-architecture#2523] [Q&A] [answered]
Suppose I had a Single Feature that works perfectly with a single shared dependency.

After a while, I decided to make more of that single feature as children of a parent.

So now I have something like:
```
        Reducer { state, action in
            parentReduce(into: &state, action: action)
        }
        .forEach(
            \.child,
             action: /Action.child
        ) {
            ChildFeature()
                .dependency(\.notSharedDependency, .init(state.something)) // 👈 the point is here, state is not available here
        }
```

So how can I access the state of each child during the `.forEach` so I can pass in the correct dependency?

Is this a good idea in the first place?

How about passing the first two arguments into the ReducerBuilder as the original [ForEach](https://developer.apple.com/documentation/swiftui/foreach) does in the SwiftUI?

I really appreciate any help you can provide.

**Comment by pyrtsa** [comment, marked as answer]
Not part of TCA yet, but this would be the prime use case for reducer readers: https://github.com/pointfreeco/swift-composable-architecture/pull/1969

In the meantime, you could implement it by manually handling the child reducer wrapping the same way (i.e. by manually calling `reduce(into:action:)` on the child reducer from the parent in a dependency injection context).

**Reply by MojtabaHs** [reply]
I see! Looking forward to updates on that PR.

By the way, When I use the `ReducerReader` for the child inside the `forEach`, the compiler breaks and raise `Ambiguous use of <something irrelevant>` 🤦🏻‍♂️

```
        Reducer { state, action in  // 👈 Compiler marks here as Ambiguous
            parentReduce(into: &state, action: action)
        }
        .forEach(
            \.child,
             action: /Action.child
        ) {
          ReducerReader { state, action in  // 👈 When using the `ReducerReader` here
            ChildFeature()
                .dependency(\.notSharedDependency, .init(state.something))
          }
        }
```

But when I use it on the parent, the compiler is satisfied but it forces me to use undesired `if let case` in the parent's state to find the proper child's state.

P.S.: I don't know why but it seems like the dependency override does not work at all and it always uses the `liveValue`, Even by passing a hard-coded value like:
```
ChildFeature()
    .dependency(\.notSharedDependency, .init("HARD-CODED STUFF HERE")
```

**Reply by pyrtsa** [reply]
If you seem to get the `liveValue` in your child reducer no matter what, one common mistake could be if you're deriving the child store using `Store.init(initialState:reducer:)` instead of `scope(state:action:)` in your views.

**Reply by MojtabaHs** [reply]
Any idea about the ambiguity?

**Reply by fabstu** [reply]
You could try changing the second state, action to different names. Not sure though. I tried putting ReducerReader inside Scope before and had issues as well, but I don't remember anymore exactly what it was.

**Reply by MojtabaHs** [reply]
Turns out is because the `ReducerReader`'s initializer and Xcode were pointing to an invalid location of the code (like always!).

So changing:
```
ReducerReader { state, action in
```

To this:

```
ReducerReader.init({ state, action in
```

Does the job and removes the ambiguity.

**Comment by MojtabaHs** [comment]
Turns out [this old drafted PR](https://github.com/pointfreeco/swift-composable-architecture/pull/1969) is exactly what I need.

Closing this discussion in favor of that.
